core/Enclave/enclave_hsm.cpp

```diff
@@ -43,767 +42,666 @@
 
-    if (plaintext == NULL || plaintext_len > EH_ENCRYPT_MAX_SIZE)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    if (cipherblob == NULL ||
-            cipherblob_len < plaintext_len + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE)
+    if (cmk == NULL || plaintext == NULL || ciphertext == NULL)
+    {
         return SGX_ERROR_INVALID_PARAMETER;
-
-    uint8_t *iv = (uint8_t *)(cipherblob + plaintext_len);
-    uint8_t *mac = (uint8_t *)(cipherblob + plaintext_len + EH_AES_GCM_IV_SIZE );
-
-    sgx_key_128bit_t dec_key;
-    ret = sgx_gcm_decrypt(&g_domain_key,
-                          &dec_key_size, (uint8_t *)&dec_key,
-                          (sgx_aes_gcm_data_ex_t *)cmk_blob);
-    if (ret != SGX_SUCCESS) {
-        printf("error(%d) unsealing key.\n", ret);
-        return ret;
     }
 
-    ret = sgx_rijndael128GCM_decrypt(&dec_key, cipherblob, plaintext_len, plaintext,
-            iv, EH_AES_GCM_IV_SIZE, aad, aad_len,
-            reinterpret_cast<uint8_t (*)[16]>(mac));
-    if (SGX_SUCCESS != ret) {
-        printf("error decrypting encrypted text\n");
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_RSA_2048:
+    case EH_RSA_3072:
+    case EH_RSA_4096:
+        ret = ehsm_rsa_encrypt(cmk, plaintext, ciphertext);
+        break;
+    case EH_SM2:
+        ret = ehsm_sm2_encrypt(cmk, plaintext, ciphertext);
+        break;
+    default:
+        return SGX_ERROR_INVALID_PARAMETER;
+        break;
     }
-
-    memset_s(&dec_key, sizeof(dec_key), 0, sizeof(dec_key));
-
     return ret;
 }
 
-sgx_status_t enclave_generate_datakey(uint32_t key_spec,
-                                  const uint8_t *cmk_blob,
-                                  size_t cmk_blob_size,
-                                  const uint8_t *context,
-                                  size_t context_len,
-                                  uint8_t *plain_key,
-                                  size_t plain_key_len,
-                                  uint8_t *encrypted_key,
-                                  size_t encrypted_key_len)
+sgx_status_t enclave_asymmetric_decrypt(const ehsm_keyblob_t *cmk, size_t cmk_size,
+                                        ehsm_data_t *ciphertext, uint32_t ciphertext_size,
+                                        ehsm_data_t *plaintext, uint32_t plaintext_size)
 {
     sgx_status_t ret = SGX_ERROR_UNEXPECTED;
-    uint8_t *datakey = NULL;
 
-    if (cmk_blob == NULL || encrypted_key ==  NULL)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    if (plain_key_len > 1024 || plain_key_len == 0)
+    if (cmk_size != APPEND_SIZE_TO_KEYBOB_T(cmk->keybloblen))
+    {
         return SGX_ERROR_INVALID_PARAMETER;
-
-    datakey = (uint8_t *)malloc(plain_key_len);
-    if (datakey == NULL) {
-        return SGX_ERROR_OUT_OF_MEMORY;
     }
 
-    ret = sgx_read_rand(datakey, plain_key_len);
-    if (ret != SGX_SUCCESS) {
-        free(datakey);
-        return ret;
+    if (cmk == NULL || plaintext == NULL || ciphertext == NULL)
```

Comment:
you need put the null check before size check, due to you will use cmk->xxxxx

Reply:
it will be changed